starknet_py/proxy_check.py

```diff
@@ -2,51 +2,72 @@
 # pyright: reportUndefinedVariable=false
 
 from abc import ABC, abstractmethod
+from typing import List, Optional
 
-from starknet_py.constants import OZ_PROXY_STORAGE_KEY
-from starknet_py.net.models import parse_address
-
+from starkware.starknet.public.abi import get_storage_var_address
 
 # noinspection PyUnresolvedReferences
 class ProxyCheck(ABC):
     @abstractmethod
-    async def is_proxy(self, contract: "Contract") -> bool:
+    async def implementation_address(self, contract: "Contract") -> Optional[int]:
         """
-        :return: `True` if contract is a proxy or `False` if it is not
+        :return: Implementation address of contract being proxied by `contract`
+            given as a parameter or None if implementation does not exist.
         """
 
     @abstractmethod
-    async def implementation_address(self, contract: "Contract") -> int:
+    async def implementation_hash(self, contract: "Contract") -> Optional[int]:
         """
-        :return: Address of contract being proxied by `contract` given as a parameter
+        :return: Implementation class hash of contract being proxied by `contract`
+            given as a parameter or None if implementation does not exist.
         """
 
 
 # noinspection PyUnresolvedReferences
 class ArgentProxyCheck(ProxyCheck):
-    async def is_proxy(self, contract: "Contract") -> bool:
-        return "get_implementation" in contract.functions
+    async def implementation_address(self, contract: "Contract") -> Optional[int]:
+        if "get_implementation" not in contract.functions:
+            return None
+
+        (result,) = await contract.functions["get_implementation"].call()
+        return result
 
-    async def implementation_address(self, contract: "Contract") -> int:
-        res = await contract.functions["get_implementation"].call()
-        return res[0]
+    async def implementation_hash(self, contract: "Contract") -> Optional[int]:
+        if "get_implementation" not in contract.functions:
+            return None
+
+        (result,) = await contract.functions["get_implementation"].call()
+        return result
 
```

Comment:
If argent is using address for proxies this will return this address and claim that it is a class_hash. We do need to check if the value is address or class_hash.